File: valid-parentheses/java/is-valid.java
// Valid Parentheses using built in Stack instead of ArrayList
    public boolean isValid(String s) {
        // init stack
        Stack<Character> stack = new Stack<>();
        // init default values with close
        HashMap<Character, Character> open_to_close = new HashMap<>();

        open_to_close.put(')', '(');
        open_to_close.put('}', '{');
        open_to_close.put(']', '[');

        // iterate over string
        for (int i = 0; i < s.length(); i++) {
            // get current char
            char current_char = s.charAt(i);
            // check if close
            if (open_to_close.containsKey(current_char)) {
                // check if stack is not empty
                if (!stack.empty()) {
                    // get last element correspondent
                    char last_element = stack.lastElement();
                    // compare last element correspondent with current value
                    char correspondent_value = open_to_close.get(current_char);
                    if (correspondent_value == last_element) {
                        // if it matched remove from stack
                        stack.pop();
                        // else return false
                    } else return false;

                    // else return false
                } else return  false;
            }
            // else add to the stack
            else  stack.add(current_char);
        }
        // check if stack empty
        return stack.empty();
    }

File: valid-parentheses/python/is-valid.py
def isValid(s):
    hasmap = {'(':')', '{':'}', '[':']' }
    stack = []
    
    # check if stack is not empty
    if len(s) == 1: return False

    # iterate over string and chec if open
    for char in s:
        # check if char is open
        if char in hasmap: # [
            # add it to the stack
            stack.append(char)
        else:
            if len(stack) == 0: return False
            # if close return  lastitem value from hashmap 
            lastItem = stack[-1] # (
            value = hasmap[lastItem] # )
            # check if equal remove from the stack
            if value != char: # value = ), char = )
                stack.append(char) 
            else: stack.pop() 
    return len(stack) == 0

# s1 = "()" # True
# print(isValid(s1))

# s2 = "()[]{}" # true
# print(isValid(s2))

# s3 = "(]" # false
# print(isValid(s3))

# s4 = "([])" # true
# print(isValid(s4))

# s5 = "["
# print(isValid(s5)) # False

# s6 = "]"

# print(isValid(s6)) # False

# s7 = "){"

# print(isValid(s7)) # False

s8 = ")(){}"

print(isValid(s8)) # False
